File: CS50/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    // chack valitity with Lung's Algo
    long card = get_long("Number: ");
    long buffer = card;
    int sum = 0;
    int sum2 = 0;
    int digits;
    bool skip = true;
    while (buffer)
    {
        int b = buffer % 10;
        if (skip)
        {
            sum2 = sum2 + b;
            skip = false;
        }
        else
        {
            b = b * 2;
            if (b >= 10)
            {
                sum = sum + b / 10 + b % 10;
            }
            else
            {
                sum = sum + b;
            }
            skip = true;
        }
        digits = digits + 1;
        buffer = buffer / 10;

    }

    int total;
    total = sum + sum2;

    // print type of card
    if (total % 10 == 0)
    {
        long power = pow(10, digits - 1);
        long power2 = pow(10, digits - 2);
        int first = card / power;
        int second = card / power2 % 10;

        if (first == 4 && digits >= 13 && digits <= 16)
        {
            printf("VISA\n");
        }
        else if (first == 3 &&
                 (second == 4 || second == 7) &&
                 digits == 15)
        {
            printf("AMEX\n");
        }
        else if (first == 5 &&
                 (second == 1 ||
                  second == 2 ||
                  second == 3 ||
                  second == 4 ||
                  second == 5) &&
                 digits == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}

File: CS50/sentimental-credit/credit.py
from cs50 import get_string

card = get_string("NUmber: ").strip()
length = len(card)

sum = 0
for i in range(length-2, -1, -2):
    # print(i)
    x = int(card[i]) * 2
    if x > 9:
        x = str(x)
        sum += int(x[0]) + int(x[1])
    else:
        sum += x

for i in range(length-1, 0, -2):
    # print(i)
    sum += int(card[i])

if length == 15 and int(card[0:2]) in [34, 37]:
    print("AMEX")
elif length >= 13 and length <= 16 and int(card[0]) == 4:
    print("VISA")
elif int(card[0:2]) in range(51, 56) and length == 16:
    print("MASTERCARD")
else:
    print("INVALID")